File: backend/src/main/java/com/sistema/ventas/services/ProductoService.java
package com.sistema.ventas.services;

import com.sistema.ventas.controllers.dto.ProductoRequest;
import com.sistema.ventas.controllers.dto.ProductoResponse;
import com.sistema.ventas.entities.Producto;
import com.sistema.ventas.repositories.ProductoRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductoService {

    private final ProductoRepository productoRepository;

    @Transactional(readOnly = true)
    public List<ProductoResponse> listarActivos() {
        return productoRepository.findByActivoTrue()
                .stream()
                .map(ProductoResponse::fromEntity)
                .toList();
    }

    @Transactional(readOnly = true)
    public ProductoResponse obtenerPorId(Long id) {
        Producto producto = productoRepository.findByIdAndActivoTrue(id)
                .orElseThrow(() -> new EntityNotFoundException("Producto no encontrado con id: " + id));
        return ProductoResponse.fromEntity(producto);
    }

    @Transactional
    public ProductoResponse crear(ProductoRequest request) {
        Producto producto = Producto.builder()
                .nombre(request.getNombre())
                .descripcion(request.getDescripcion())
                .precio(request.getPrecio())
                .stock(request.getStock())
                .build();
        return ProductoResponse.fromEntity(productoRepository.save(producto));
    }

    @Transactional
    public ProductoResponse actualizar(Long id, ProductoRequest request) {
        Producto producto = productoRepository.findByIdAndActivoTrue(id)
                .orElseThrow(() -> new EntityNotFoundException("Producto no encontrado con id: " + id));
        producto.setNombre(request.getNombre());
        producto.setDescripcion(request.getDescripcion());
        producto.setPrecio(request.getPrecio());
        producto.setStock(request.getStock());
        return ProductoResponse.fromEntity(productoRepository.save(producto));
    }

    @Transactional
    public void eliminar(Long id) {
        Producto producto = productoRepository.findByIdAndActivoTrue(id)
                .orElseThrow(() -> new EntityNotFoundException("Producto no encontrado con id: " + id));
        producto.setActivo(false);
        productoRepository.save(producto);
    }
}

File: frontend/src/services/productoService.ts
import { api } from './api'

export interface Producto {
  id: number
  nombre: string
  descripcion: string | null
  precio: number
  stock: number
  activo: boolean
  createdAt: string
  updatedAt: string
}

export interface ProductoRequest {
  nombre: string
  descripcion?: string
  precio: number
  stock: number
}

const API_URL = '/api/productos'

export const productoService = {
  async listar(): Promise<Producto[]> {
    const { data } = await api.get<Producto[]>(API_URL)
    return data
  },

  async obtener(id: number): Promise<Producto> {
    const { data } = await api.get<Producto>(`${API_URL}/${id}`)
    return data
  },

  async crear(request: ProductoRequest): Promise<Producto> {
    const { data } = await api.post<Producto>(API_URL, request)
    return data
  },

  async actualizar(id: number, request: ProductoRequest): Promise<Producto> {
    const { data } = await api.put<Producto>(`${API_URL}/${id}`, request)
    return data
  },

  async eliminar(id: number): Promise<void> {
    await api.delete(`${API_URL}/${id}`)
  },
}
